modules/GeneticsPortal.py

```diff
@@ -1,528 +1,261 @@
-    parser.add_argument('--schemaFile', help='OpenTargets JSON schema file (Draft-4 compatible!!).', type=str)
-    parser.add_argument('--cores', help='Number of computing cores available for the evidence generation.', type=int, default=2)
-    parser.add_argument('--outputFile', help='Name of the gzipped json output file.', type=str, default='output.json.gz')
-    parser.add_argument('--sample', help='If provided this many randomly selected of evidences will be generated.', type=int, required=False)
-    parser.add_argument('--threshold', help='If provided, evidences will be filtered for locus to gene score at the defined threshold.', type=float, required=False)
-    args = parser.parse_args()
-
-    # Parse input parameters:
-    inputFile = args.inputFile
-    cores = args.cores
-    outputFile = args.outputFile
-    sample = args.sample
-    schemaFile = args.schemaFile
-    threshold = args.threshold
-
-    #  Opening input file parquet of tsv:
-    if 'parquet' in inputFile:
-        genetics_dataframe = pd.read_parquet(inputFile)
-    elif '.tsv' in inputFile:
-        genetics_dataframe = pd.read_csv(inputFile, sep='\t')
+    # Load association statistics (only pvalue is required) from top loci table
+    pvals = (
+        spark.read.parquet(in_toploci)
+        # # Calculate pvalue from the mantissa and exponent
+        # .withColumn('pval', col('pval_mantissa') * pow(10, col('pval_exponent')))
+        # # NB. be careful as very small floats will be set to 0, we can se these
+        # # to the smallest possible float instead
+        # .withColumn('pval',
+        #     when(col('pval') == 0, sys.float_info.min)
+        #     .otherwise(col('pval'))
+        # )
+        # Keep required fields
+        .select('study_id', 'chrom', 'pos', 'ref', 'alt', 
+            'pval_mantissa', 'pval_exponent','odds_ratio','oddsr_ci_lower', 'oddsr_ci_upper')
+    )
 
-    logging.info('Loading data completed.')
+    # Load (a) disease information, (b) sample size from the study table
+    study_info = (
+        spark.read.parquet(in_study)
+        .select(
+            'study_id', 'pmid', 'pub_date', 'pub_author', 'trait_reported',
+            'trait_efos',
+            col('n_initial').alias('sample_size') # Rename to sample size
+        )
+                
+        # Warning! Not all studies have an EFO annotated. Also, some have
+        # multiple EFOs! We need to decide a strategy to deal with these.
+        
+        # # For example, only keep studies with 1 efo:
+        # .filter(size(col('trait_efos')) == 1)
+        # .withColumn('efo', col('trait_efos').getItem(0))
+        # .drop('trait_efos')
+
+        # Or, drop rows with no EFO and then explode array to multiple rows
+        .withColumn('trait_efos', when(col('trait_efos').isNotNull(),
+                                       expr('filter(trait_efos, t -> length(t) > 0)')))
```

Comment:
I would omit this piece of code. This new column function filters those arrays in trait_efos with a size of >= 0 whenever the value is not null.
What is the difference with line 147? There are 753 studies with an empty array and we're not interested in such cases.

Reply:
It took a while for me to figure out what's going on here (first of all, you are right, row 147 is not needed). So we need to apply an extra filter, as there are 12 studies with empty string as efo. Where the explosion would still generated rows with no efo. You can try it out. Anyway, I'll simplify the code as much as I can and do an other commit.

```
+--------------------+--------------------+-----------+------------+
|            study_id|      trait_reported|sample_size|         efo|
+--------------------+--------------------+-----------+------------+
|          GCST008760|Perceived sweetne...|        686|            |
|          GCST008311|Lymphangioleiomyo...|       1278|            |
|          GCST008870|Keratinocyte canc...|     358840|            |
|     NEALE2_3761_raw|Age hay fever, rh...|      72232|            |
|          GCST008765|Perceived intensi...|       1757|            |
|          GCST008766|Perceived intensi...|       1757|            |
|          GCST009418|Lactose tolerance...|         60|            |
|          GCST008767|Perceived intensi...|        686|            |
|          GCST008763|Perceived intensi...|       1757|            |
|          GCST008768|Perceived intensi...|       2443|            |
|          GCST008769|Perceived intensi...|       1757|            |
|          GCST008764|Perceived intensi...|       1757|            |
```

Reply:
I don't think I'm understanding this correctly: by keeping this piece of code and not the previous one:
```
.filter(col('trait_efos').isNotNull() & (size(col('trait_efos')) > 0))
```
you are keeping those rows where the `trait_efos` array is empty. So the remaining code contradicts with your comment above. Can you explain where I'm getting things wrong?

Reply:
So in my understanding that expression filters for rows where the `trait_efos` value is not null and the length of the value is greater than zero, filtering out values where the only element is an empty string eg:
```
[ "" ]
```